pystac/link.py

```diff
@@ -298,6 +298,10 @@ def resolve_stac_object(self, root: Optional["Catalog_Type"] = None) -> "Link":
                     if self.owner is not None:
                         if isinstance(self.owner, pystac.Catalog):
                             stac_io = self.owner._stac_io
+                        else:
+                            owner_root = self.owner.get_root()
+                            if owner_root is not None:
+                                stac_io = owner_root._stac_io
```

Comment:
Recursion issue caused by trying to check root, which will try to resolve the root link, happens when the root link is the one currently being resolved. Avoided with:

```suggestion
                        elif self.rel != pystac.RelType.ROOT:
                            owner_root = self.owner.get_root()
                            if owner_root is not None:
                                stac_io = owner_root._stac_io
```